apps/src/lib/node/ledger/shell/prepare_proposal/tx_bins.rs

```diff
@@ -0,0 +1,355 @@
+//! Primitives that facilitate keeping track of the number
+//! of bytes utilized by some Tendermint consensus round's proposal.
+//!
+//! This is important, because Tendermint places an upper bound
+//! on the size of a block, rejecting blocks whose size exceeds
+//! the limit stated in [`RequestPrepareProposal`].
+//!
+//! In the current implementation, each kind of transaction in
+//! Namada gets a portion of (i.e. threshold over) the total
+//! allotted space.
+
+// TODO: what if a tx has a size greater than the threshold for
+// its bin? how do we handle this? if we keep it in the mempool
+// forever, it'll be a DoS vec, as we can make nodes run out of
+// memory! maybe we should allow block decisions for txs that are
+// too big to fit in their respective bin? in these special block
+// decisions, we would only decide proposals with "large" txs
+
+// TODO: refactor our measure of space to also reflect gas costs.
+// the total gas of all chosen txs cannot exceed the configured max
+// gas per block, otherwise a proposal will be rejected!
+
+use num_rational::Ratio;
+
+use crate::facade::tendermint_proto::abci::RequestPrepareProposal;
+
+/// Allotted space for a batch of transactions in some proposed block,
+/// measured in bytes.
+///
+/// We keep track of the current space utilized by:
+///
+///   - Protocol transactions.
+///   - DKG decrypted transactions.
+///   - DKG encrypted transactions.
+#[derive(Default)]
+pub struct TxAllottedSpace {
+    /// The total space Tendermint has allotted to the
+    /// application for the current block height.
+    bytes_provided_by_tendermint: u64,
+    /// The current space utilized by protocol transactions.
+    protocol_txs: TxBin,
+    /// The current space utilized by DKG encrypted transactions.
+    encrypted_txs: TxBin,
+    /// The current space utilized by DKG decrypted transactions.
+    decrypted_txs: TxBin,
+}
+
+impl From<&RequestPrepareProposal> for TxAllottedSpace {
+    #[inline]
+    fn from(req: &RequestPrepareProposal) -> Self {
+        let tendermint_max_block_space_in_bytes = req.max_tx_bytes as u64;
+        Self::init(tendermint_max_block_space_in_bytes)
+    }
+}
+
+impl TxAllottedSpace {
+    /// Construct a new [`TxAllottedSpace`], with an upper bound
+    /// on the max number of txs in a block defined by Tendermint.
+    #[inline]
+    pub fn init(tendermint_max_block_space_in_bytes: u64) -> Self {
+        let max = tendermint_max_block_space_in_bytes;
+        let mut bins = Self {
+            bytes_provided_by_tendermint: max,
```

Comment:
Is this meant to represent `max_tx_bytes` as from https://github.com/tendermint/tendermint/blob/v0.37.x/spec/abci/abci++_methods.md#parameters-and-types ? I think we should name it something indicating that as `bytes_provided_by_tendermint` sounds like it could be the `request.txs.len()` (which is possible to be more than `request.max_tx_bytes` it seems). Sorry to nit again, even `max_tx_bytes` or `max_block_space_in_bytes` seems like an appropriate field name